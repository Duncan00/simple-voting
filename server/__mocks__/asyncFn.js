'use strict';

const util = require('util');

const asyncFn = fname =>
	jest.fn(async (...args) => {
		const msg = `unexpected call to ${fname} with (${util.inspect(
			{args},
			10
		)}), did you miss any mockImplementationOnce?`;
		throw new Error(msg);
	});

module.exports = asyncFn;
