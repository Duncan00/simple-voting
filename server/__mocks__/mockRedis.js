'use strict';

const Redis = require('ioredis-mock');
const redis = new Redis();
// const asyncFn = require('./asyncFn');
//
// const redis = {
// 	mock_object_name: 'mock-redis-object',
// 	incr: asyncFn('redis#incr'),
// 	get: asyncFn('redis#get'),
// 	hmset: asyncFn('redis#hmset'),
// 	hgetall: asyncFn('redis#hgetall'),
// 	sismember: asyncFn('redis#sismember'),
// 	sadd: asyncFn('redis#sadd'),
// 	hincrby: asyncFn('redis#hincrby'),
// 	ping: asyncFn('redis#ping'),
// 	on: asyncFn('redis#on')
// };
const getRedis = jest.fn().mockResolvedValue(redis);
jest.doMock('../services/redis/getRedis', () => getRedis);

module.exports = redis;
