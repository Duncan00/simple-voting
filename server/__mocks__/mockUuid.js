'use strict';

// make uuid predictable in snapshots
jest.mock('uuid/v4', () => {
	let count = 0;
	return () => `uuid-${count++}`;
});
