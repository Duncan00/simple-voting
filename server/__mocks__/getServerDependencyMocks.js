'use strict';

/**
 *  Setup mocks for system boundaries (third party, data storage, etc) for easier testing
 */

const mockRedis = require('./mockRedis');

require('./mockUuid');

module.exports = {
	redis: mockRedis
};
