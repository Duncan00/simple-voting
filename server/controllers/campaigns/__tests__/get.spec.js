'use strict';

const moment = require('moment');
const supertest = require('supertest');
const _ = require('lodash');
const mockdate = require('mockdate');
const initServer = () => require('../../../initServer')();
const getServerDependencyMocks = () =>
	require('../../../__mocks__/getServerDependencyMocks');
const hashHkid = require('../hashHkid');

describe('GET /v1/campaigns', () => {
	let server;

	const campaign_records = [
		{
			id: '1a044192-db7e-4fd3-9334-c9344360286b',
			title: 'Current campaign 1',
			start_date: '2019-01-01',
			end_date: '2019-01-31',
			'number_of_votes:Michael Jordan': '0',
			'number_of_votes:Tim Duncan': '99'
		},

		{
			id: '2a044192-db7e-4fd3-9334-c9344360286b',
			title: 'Current campaign 2',
			start_date: '2019-01-01',
			end_date: '2019-01-31',
			'number_of_votes:Michael Jordan': '100',
			'number_of_votes:Tim Duncan': '0'
		},

		{
			id: '3a044192-db7e-4fd3-9334-c9344360286b',
			title: 'Current campaign 3',
			start_date: '2019-01-01',
			end_date: '2019-01-31',
			'number_of_votes:Michael Jordan': '50',
			'number_of_votes:Tim Duncan': '51'
		},

		{
			id: 'aa044192-db7e-4fd3-9334-c9344360286b',
			title: 'Old campaign 1',
			start_date: '2018-11-01',
			end_date: '2018-11-30',
			'number_of_votes:Michael Jordan': '100000',
			'number_of_votes:Tim Duncan': '0'
		},

		{
			id: 'ba044192-db7e-4fd3-9334-c9344360286b',
			title: 'Old campaign 2',
			start_date: '2018-12-01',
			end_date: '2018-12-31',
			'number_of_votes:Michael Jordan': '200',
			'number_of_votes:Tim Duncan': '0'
		},

		{
			id: 'fa044192-db7e-4fd3-9334-c9344360286b',
			title: 'Future campaign',
			start_date: '2019-02-01',
			end_date: '2019-02-28',
			'number_of_votes:Michael Jordan': '0',
			'number_of_votes:Tim Duncan': '0'
		}
	];

	const hkid = 'A123456(3)';
	const hashed_hkid = hashHkid(hkid);

	beforeEach(async () => {
		jest.resetModules();
		mockdate.set('2019-01-01');
		const {redis} = getServerDependencyMocks();
		await Promise.all(
			campaign_records.map(async campaign_record => {
				return Promise.all([
					redis.hmset(
						`campaign:{${campaign_record.id}}`,
						campaign_record
					),
					redis.zadd(
						'campaign_end_date',
						moment
							.utc(campaign_record.end_date)
							.endOf('day')
							.format('X'),
						campaign_record.id
					)
				]);
			})
		);
		// Voted "Michael Jordan" at "Current campaign 2"
		await redis.sadd(
			`voted:{2a044192-db7e-4fd3-9334-c9344360286b}`,
			hashed_hkid
		);
		await redis.sadd(
			`voted:{2a044192-db7e-4fd3-9334-c9344360286b}:Michael Jordan`,
			hashed_hkid
		);

		server = await initServer();
	});

	afterEach(() => {
		server.close();
		mockdate.reset();
	});

	// TODO: Different combination of test cases should be added
	describe('When query without query strings', () => {
		test('200 - Responds a list of all campaigns started', async () => {
			const {body, status} = await supertest(server).get(`/v1/campaigns`);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(200);
		});
	});

	describe('When query with hkid', () => {
		test('200 - Responds a list of all campaigns started', async () => {
			const {body, status} = await supertest(server).get(
				`/v1/campaigns?hkid=${hkid}`
			);

			expect(body).toMatchSnapshot();
			expect(status).toEqual(200);
		});
	});
});
