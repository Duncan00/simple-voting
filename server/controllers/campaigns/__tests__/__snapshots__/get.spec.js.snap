// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /v1/campaigns When query 200 - Responds a list of all campaigns started 1`] = `
Object {
  "data": Object {
    "campaigns": Array [
      Object {
        "candidates": Array [
          Object {
            "name": "Michael Jordan",
            "number_of_votes": 50,
            "voted": false,
          },
          Object {
            "name": "Tim Duncan",
            "number_of_votes": 51,
            "voted": false,
          },
        ],
        "end_date": "2019-01-31",
        "id": "3a044192-db7e-4fd3-9334-c9344360286b",
        "start_date": "2019-01-01",
        "title": "Current campaign 3",
      },
      Object {
        "candidates": Array [
          Object {
            "name": "Michael Jordan",
            "number_of_votes": 100,
            "voted": false,
          },
          Object {
            "name": "Tim Duncan",
            "number_of_votes": 0,
            "voted": false,
          },
        ],
        "end_date": "2019-01-31",
        "id": "2a044192-db7e-4fd3-9334-c9344360286b",
        "start_date": "2019-01-01",
        "title": "Current campaign 2",
      },
      Object {
        "candidates": Array [
          Object {
            "name": "Michael Jordan",
            "number_of_votes": 0,
            "voted": false,
          },
          Object {
            "name": "Tim Duncan",
            "number_of_votes": 99,
            "voted": false,
          },
        ],
        "end_date": "2019-01-31",
        "id": "1a044192-db7e-4fd3-9334-c9344360286b",
        "start_date": "2019-01-01",
        "title": "Current campaign 1",
      },
      Object {
        "candidates": Array [
          Object {
            "name": "Michael Jordan",
            "number_of_votes": 200,
            "voted": false,
          },
          Object {
            "name": "Tim Duncan",
            "number_of_votes": 0,
            "voted": false,
          },
        ],
        "end_date": "2018-12-31",
        "id": "ba044192-db7e-4fd3-9334-c9344360286b",
        "start_date": "2018-12-01",
        "title": "Old campaign 2",
      },
      Object {
        "candidates": Array [
          Object {
            "name": "Michael Jordan",
            "number_of_votes": 100000,
            "voted": false,
          },
          Object {
            "name": "Tim Duncan",
            "number_of_votes": 0,
            "voted": false,
          },
        ],
        "end_date": "2018-11-30",
        "id": "aa044192-db7e-4fd3-9334-c9344360286b",
        "start_date": "2018-11-01",
        "title": "Old campaign 1",
      },
    ],
  },
  "meta": Object {
    "code": 20000,
    "message": "OK",
  },
}
`;
